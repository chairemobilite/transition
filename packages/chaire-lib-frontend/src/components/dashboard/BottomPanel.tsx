/*
 * Copyright 2022, Polytechnique Montreal and contributors
 *
 * This file is licensed under the MIT License.
 * License text available at https://opensource.org/licenses/MIT
 */
import React from 'react';

import { Contribution, LayoutSectionProps } from '../../services/dashboard/DashboardContribution';

interface BottomPanelProps extends LayoutSectionProps {
    activeSection: string;
    contributions: Contribution<LayoutSectionProps>[];
}

/**
 * Bottom panel widget displaying the contributions in parameter. This class is
 * not to be extended. @see Contribution for how to provide widgets to display
 * in the bottom panel
 * @param param0
 * @returns
 */
const BottomPanel: React.FunctionComponent<BottomPanelProps> = ({ contributions, ...props }: BottomPanelProps) => {
    const contributionElements = React.useMemo(
        () =>
            contributions
                .filter((contrib) => contrib.section === undefined || contrib.section === props.activeSection)
                .map((contrib) => contrib.create({ ...props, key: `bottomPanelEl${contrib.id}` })),
        [props.activeSection]
    );

    return <section id="tr__bottom-panel">{contributionElements}</section>;
};

export default BottomPanel;
