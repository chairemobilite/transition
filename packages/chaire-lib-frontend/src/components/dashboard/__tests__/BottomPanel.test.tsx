/*
 * Copyright 2022, Polytechnique Montreal and contributors
 *
 * This file is licensed under the MIT License.
 * License text available at https://opensource.org/licenses/MIT
 */
import React from 'react';
import { render } from '@testing-library/react';

import BottomPanel from '../BottomPanel';

const contributions = [
    {
        id: 'testBottomPanel',
        section: 'test',
        placement: 'bottomPanel' as const,
        create: (_props: any) => <div>test</div>
    },
    {
        id: 'mainBottomPanel',
        placement: 'bottomPanel' as const,
        create: (_props: any) => <div>Should always be there</div>
    },
    {
        id: 'fooBottomPanel',
        section: 'foo',
        placement: 'bottomPanel' as const,
        create: (props: any) => <div {...props}>Foo section, with props</div>
    }
];

test('One active section with contribution', () => {
    const { container } = render(<BottomPanel
        activeSection = 'test'
        contributions = {contributions}
    />);
    expect(container).toMatchSnapshot();
});

test('One active section with contribution and props', () => {
    const { container } = render(<BottomPanel
        activeSection = 'foo'
        contributions = {contributions}
    />);
    expect(container).toMatchSnapshot();
});

test('active section without contribution', () => {
    const { container } = render(<BottomPanel
        activeSection = 'bar'
        contributions = {contributions}
    />);
    expect(container).toMatchSnapshot();
});

test('Active section change', () => {
    const { rerender, container } = render(<BottomPanel
        activeSection = 'test'
        contributions = {contributions}
    />);
    expect(container).toMatchSnapshot();
    rerender(<BottomPanel
        activeSection = 'other' // there is no contributions for activeSection other, so it should not display the activeSection
        contributions = {contributions}
    />);
    expect(container).toMatchSnapshot();
});
