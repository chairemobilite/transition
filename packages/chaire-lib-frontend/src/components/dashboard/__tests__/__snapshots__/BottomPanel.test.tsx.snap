// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Active section change 1`] = `
<div>
  <section
    id="tr__bottom-panel"
  >
    <div>
      test
    </div>
    <div>
      Should always be there
    </div>
  </section>
</div>
`;

exports[`Active section change 2`] = `
<div>
  <section
    id="tr__bottom-panel"
  >
    <div>
      Should always be there
    </div>
  </section>
</div>
`;

exports[`One active section with contribution 1`] = `
<div>
  <section
    id="tr__bottom-panel"
  >
    <div>
      test
    </div>
    <div>
      Should always be there
    </div>
  </section>
</div>
`;

exports[`One active section with contribution and props 1`] = `
<div>
  <section
    id="tr__bottom-panel"
  >
    <div>
      Should always be there
    </div>
    <div
      activesection="foo"
    >
      Foo section, with props
    </div>
  </section>
</div>
`;

exports[`active section without contribution 1`] = `
<div>
  <section
    id="tr__bottom-panel"
  >
    <div>
      Should always be there
    </div>
  </section>
</div>
`;
